spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/RewritePositionDeletesSparkAction.java

```diff
@@ -0,0 +1,501 @@
+    RewritePositionDeletesCommitManager.CommitService commitService =
+        commitManager.service(groupsPerCommit);
+    commitService.start();
+
+    // Start rewrite tasks
+    Tasks.foreach(groupStream)
+        .suppressFailureWhenFinished()
+        .executeWith(rewriteService)
+        .noRetry()
+        .onFailure(
+            (fileGroup, exception) ->
+                LOG.error("Failure during rewrite group {}", fileGroup.info(), exception))
+        .run(fileGroup -> commitService.offer(rewriteDeleteFiles(ctx, fileGroup)));
+    rewriteService.shutdown();
+
+    // Stop Commit service
+    commitService.close();
+    List<RewritePositionDeleteGroup> commitResults = commitService.results();
+    if (commitResults.size() == 0) {
+      LOG.error(
+          "{} is true but no rewrite commits succeeded. Check the logs to determine why the individual "
+              + "commits failed. If this is persistent it may help to increase {} which will break the rewrite operation "
+              + "into smaller commits.",
+          PARTIAL_PROGRESS_ENABLED,
+          PARTIAL_PROGRESS_MAX_COMMITS);
+    }
+
+    List<PositionDeleteGroupRewriteResult> rewriteResults =
+        commitResults.stream()
+            .map(RewritePositionDeleteGroup::asResult)
+            .collect(Collectors.toList());
+    return ImmutableResult.builder().rewriteResults(rewriteResults).build();
+  }
+
+  Stream<RewritePositionDeleteGroup> toGroupStream(
+      RewriteExecutionContext ctx,
+      Map<StructLike, List<List<PositionDeletesScanTask>>> groupsByPartition) {
+    Stream<RewritePositionDeleteGroup> rewriteFileGroupStream =
+        groupsByPartition.entrySet().stream()
+            .flatMap(
+                e -> {
+                  StructLike partition = e.getKey();
+                  List<List<PositionDeletesScanTask>> scanGroups = e.getValue();
+                  return scanGroups.stream()
+                      .map(
+                          tasks -> {
+                            int globalIndex = ctx.currentGlobalIndex();
+                            int partitionIndex = ctx.currentPartitionIndex(partition);
+                            // as coerced partition used for map, return original partition
+                            PositionDeleteGroupInfo info =
+                                ImmutablePositionDeleteGroupInfo.builder()
+                                    .globalIndex(globalIndex)
+                                    .partitionIndex(partitionIndex)
+                                    .partition(partition)
+                                    .build();
+                            return new RewritePositionDeleteGroup(info, tasks);
+                          });
+                });
+
+    return rewriteFileGroupStream.sorted(rewriteGroupComparator());
+  }
+
+  private Comparator<RewritePositionDeleteGroup> rewriteGroupComparator() {
```

Comment:
What about having a method in `RewritePositionDeleteGroup` that accepts `RewriteJobOrder` and returns a comparator to simplify the action, which is already complicated?